src/core/qgsdbfilterproxymodel.cpp

```diff
@@ -17,24 +17,68 @@
 
 #include "qgsdbfilterproxymodel.h"
 
+#include <QStandardItemModel>
+
 QgsDatabaseFilterProxyModel::QgsDatabaseFilterProxyModel( QObject *parent ): QSortFilterProxyModel( parent )
 {
 
 }
 
-bool QgsDatabaseFilterProxyModel::filterAcceptsRow( int row, const QModelIndex &source_parent ) const
+bool QgsDatabaseFilterProxyModel::filterAcceptsRow( int source_row, const QModelIndex &source_parent ) const
 {
-  //if parent is valid, we have a toplevel item that should be always shown
-  if ( !source_parent.isValid() )
+  if ( filterAcceptsRowItself( source_row, source_parent ) )
+    return true;
+
+//accept if any of the parents is accepted on it's own merits
+  QModelIndex parent = source_parent;
+  while ( parent.isValid() )
+  {
+    if ( filterAcceptsRowItself( parent.row(), parent.parent() ) )
+      return true;
+    parent = parent.parent();
+  }
+
+  //accept if any of the children is accepted on it's own merits
+  if ( hasAcceptedChildren( source_row, source_parent ) )
```

Comment:
We could use https://doc.qt.io/qt-5/qsortfilterproxymodel.html#recursiveFilteringEnabled-prop and avoid all this manual code (we should use this throughout qgis now, we could drop lots of code as a result)

Reply:
good catch, we can even skip the re-implemtation of QSortFilterProxyModel